python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -542,42 +584,339 @@ def asset_key_for_input(self, input_name: str) -> AssetKey:
+        you can use asset_partition_key_range_for_output to get all of the partitions being materialized
+        by the backfill.
+
+        Args:
+            output_name (str): The name of the output produced by the asset to get the key range for.
+                If not provided for assets defined with the `@asset` decorator, the name of the output
+                will be automatically provided.
+
+        Examples:
+            .. code-block:: python
+
+                partitions_def = DailyPartitionsDefinition("2023-08-20")
+
+                @asset(
+                    partitions_def=partitions_def
+                )
+                def an_asset(context: AssetExecutionContext):
+                    context.log.info(context.asset_partition_key_range_for_output())
+
+
+                # running a backfill of the 2023-08-21 through 2023-08-25 partitions of this asset will log:
+                #   PartitionKeyRange(start="2023-08-21", end="2023-08-25")
+
+                @multi_asset(
+                    outs={
+                        "first_asset": AssetOut(key=["my_assets", "first_asset"]),
+                        "second_asset": AssetOut(key=["my_assets", "second_asset"])
+                    }
+                    partitions_def=partitions_def,
+                )
+                def a_multi_asset(context: AssetExecutionContext):
+                    context.log.info(context.asset_partition_key_range_for_output("first_asset"))
+                    context.log.info(context.asset_partition_key_range_for_output("second_asset"))
+
+
+                # running a backfill of the 2023-08-21 through 2023-08-25 partitions of this asset will log:
+                #   PartitionKeyRange(start="2023-08-21", end="2023-08-25")
+                #   PartitionKeyRange(start="2023-08-21", end="2023-08-25")
+
+
+                @asset(
+                    partitions_def=partitions_def,
+                    ins={
+                        "self_dependent_asset": AssetIn(partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1))
+                    }
+                )
+                def self_dependent_asset(context: AssetExecutionContext, self_dependent_asset):
+                    context.log.info(context.asset_partition_key_range_for_output())
+
+                # running a backfill of the 2023-08-21 through 2023-08-25 partitions of this asset will log:
+                #   PartitionKeyRange(start="2023-08-21", end="2023-08-25")
+
+        """
         return self._step_execution_context.asset_partition_key_range_for_output(output_name)
 
     @public
     def asset_partition_key_range_for_input(self, input_name: str) -> PartitionKeyRange:
-        """Return the PartitionKeyRange for the corresponding input. Errors if there is more or less than one."""
+        """Return the PartitionKeyRange for the corresponding input. Errors if there is more or less than one.
```

Comment:
I'd like to elaborate on when there would be more or less than one key range, but i'm not sure what those conditions are

Reply:
It's whenever a downstream asset depends on multiple partitions that aren't in one contiguous chunk.

I.e. Upstream has static partitions def `[a, b, c]` and downstream is unpartitioned, but depends on partitions a and c. The upstream partition key ranges would then be `[a, a]` and `[c, c]` causing this function to error.